File: src/examples/cpp/protobuf_channel/module/normal_subscriber_module/normal_subscriber_module.cc
#include "normal_subscriber_module/normal_subscriber_module.h"
#include "aimrt_module_protobuf_interface/channel/protobuf_channel.h"
#include "aimrt_module_protobuf_interface/util/protobuf_tools.h"

#include "yaml-cpp/yaml.h"

namespace aimrt::examples::cpp::protobuf_channel::normal_subscriber_module {

bool NormalSubscriberModule::Initialize(aimrt::CoreRef core) noexcept {
  core_ = core;

  try {
    // Read cfg
    const auto configurator = core_.GetConfigurator();
    if (configurator) {
      YAML::Node cfg_node = YAML::LoadFile(std::string(configurator.GetConfigFilePath()));
      topic_name_ = cfg_node["topic_name"].as<std::string>();
    }

    // 订阅事件
    subscriber_ = core_.GetChannelHandle().GetSubscriber(topic_name_);
    AIMRT_CHECK_ERROR_THROW(subscriber_, "Get subscriber for topic '{}' failed.", topic_name_);

    bool ret = aimrt::channel::SubscribeCo<aimrt::protocols::example::ExampleEventMsg>(
        subscriber_,
        std::bind(&NormalSubscriberModule::EventHandle, this, std::placeholders::_1));
    AIMRT_CHECK_ERROR_THROW(ret, "Subscribe failed.");

  } catch (const std::exception& e) {
    AIMRT_ERROR("Init failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Init succeeded.");

  return true;
}

bool NormalSubscriberModule::Start() noexcept { return true; }

void NormalSubscriberModule::Shutdown() noexcept {}

co::Task<void> NormalSubscriberModule::EventHandle(const aimrt::protocols::example::ExampleEventMsg& data) {
  AIMRT_INFO("Receive new pb event, data: {}", aimrt::Pb2CompactJson(data));

  co_return;
}

}  // namespace aimrt::examples::cpp::protobuf_channel::normal_subscriber_module

File: src/examples/py/protobuf_channel/py/normal_subscriber_module.py
import aimrt_py
import yaml

from google.protobuf.json_format import MessageToJson
import event_pb2


class NormalSubscriberModule(aimrt_py.ModuleBase):
    def __init__(self):
        super().__init__()
        self.core = aimrt_py.CoreRef()
        self.logger = aimrt_py.LoggerRef()

        self.topic_name = "test_topic"
        self.subscriber = aimrt_py.SubscriberRef()

    def Info(self):
        info = aimrt_py.ModuleInfo()
        info.name = "NormalSubscriberModule"
        return info

    def Initialize(self, core):
        self.core = core
        self.logger = self.core.GetLogger()

        # log
        aimrt_py.info(self.logger, "Module initialize")

        try:
            # configure
            configurator = self.core.GetConfigurator()
            if (configurator):
                module_cfg_file_path = configurator.GetConfigFilePath()
                if (module_cfg_file_path):
                    with open(module_cfg_file_path, 'r') as file:
                        data = yaml.safe_load(file)
                        self.topic_name = str(data["topic_name"])

            # channel-subscriber
            self.subscriber = self.core.GetChannelHandle().GetSubscriber(self.topic_name)
            if (not self.subscriber):
                aimrt_py.error(self.logger, "Get subscriber for '{}' failed.".format(self.topic_name))
                return False

            def EventHandle(msg):
                aimrt_py.info(self.logger, "Receive new pb event, data: {}".format(MessageToJson(msg)))

            aimrt_py.Subscribe(self.subscriber, event_pb2.ExampleEventMsg, EventHandle)

        except Exception as e:
            aimrt_py.error(self.logger, "Initialize failed. {}".format(e))
            return False

        return True

    def Start(self):
        aimrt_py.info(self.logger, "Module start")

        return True

    def Shutdown(self):
        aimrt_py.info(self.logger, "Module shutdown")
